mlir/lib/Conversion/ConvertToAIRPass.cpp

```diff
@@ -2318,6 +2318,267 @@ class ScfForallToLaunchConversion : public OpRewritePattern<scf::ForallOp> {
   // bool generateSegment;
 };
 
+/// Build a strided memref type by applying `permutationMap` tp `memRefType`.
+static MemRefType inferTransposeResultType(MemRefType memRefType,
+                                           AffineMap permutationMap) {
+  auto rank = memRefType.getRank();
+  auto originalSizes = memRefType.getShape();
+  auto [originalStrides, offset] = getStridesAndOffset(memRefType);
+  assert(originalStrides.size() == static_cast<unsigned>(rank));
+
+  // Compute permuted sizes and strides.
+  SmallVector<int64_t> sizes(rank, 0);
+  SmallVector<int64_t> strides(rank, 1);
+  for (const auto &en : llvm::enumerate(permutationMap.getResults())) {
+    unsigned position = cast<AffineDimExpr>(en.value()).getPosition();
+    sizes[en.index()] = originalSizes[position];
+    strides[en.index()] = originalStrides[position];
+  }
+
+  return MemRefType::Builder(memRefType)
+      .setShape(sizes)
+      .setLayout(
+          StridedLayoutAttr::get(memRefType.getContext(), offset, strides));
+}
+
+static void extractStridesFromMemrefType(MemRefType memrefTy,
+                                         OpBuilder &builder,
+                                         SmallVector<Value, 4> &strides) {
+  // get the strides and offsets from the memref type
+  int64_t offset;
+  SmallVector<int64_t, 4> layout_strides;
+  auto successStrides = getStridesAndOffset(memrefTy, layout_strides, offset);
+  if (failed(successStrides)) {
+    llvm::outs() << "Failed to get strides\n";
+    return;
+  }
+
+  for (auto s : layout_strides)
+    strides.push_back(
+        builder.create<arith::ConstantIndexOp>(builder.getUnknownLoc(), s));
+}
+
+static void extractSizesFromMemrefType(MemRefType memrefTy, OpBuilder &builder,
+                                       SmallVector<Value, 4> &sizes) {
+  for (auto s : memrefTy.getShape())
+    sizes.push_back(
+        builder.create<arith::ConstantIndexOp>(builder.getUnknownLoc(), s));
+}
+
+static void extractOffsetsFromSubview(memref::SubViewOp subview,
+                                      OpBuilder &builder,
+                                      SmallVector<Value, 4> &offsets) {
+  auto subview_offsets = subview.getOffsets().begin();
+  auto static_offsets = subview.getStaticOffsets();
+  auto loc = subview.getLoc();
+
+  for (auto o : static_offsets) {
+    if (o >= 0)
```

Comment:
Are you checking for the ShapedType::kDynamic value here? https://mlir.llvm.org/docs/Dialects/MemRef/#memrefsubview-memrefsubviewop
Please add a comment in the code :)